pkg/codegen/pcl/binder.go

```diff
@@ -470,6 +498,188 @@ func getBooleanAttributeValue(attr *model.Attribute) (bool, *hcl.Diagnostic) {
+	attributes map[string]hclsyntax.Expression,
+) (*schema.ParameterizationDescriptor, *hcl.Diagnostic) {
+	descriptor := &schema.ParameterizationDescriptor{}
+
+	// read name, version, and value attributes
+	for key, value := range attributes {
+		switch key {
+		case "name":
+			name, err := evaluateLiteralExpr(value)
+			if err != nil {
+				return nil, errorf(value.Range(), "invalid name attribute of parameterization of %q: %v", packageName, err)
+			}
+			if name == "" {
+				return nil, errorf(value.Range(), "name attribute of parameterization of %q cannot be empty", packageName)
+			}
+			descriptor.Name = name
+		case "version":
+			version, _ := evaluateLiteralExpr(value)
+			parsedVersion, err := semver.Make(version)
+			if err != nil {
+				return nil, errorf(value.Range(),
+					"invalid version %q of parameterization for package %q: %v", version, packageName, err)
+			}
+			descriptor.Version = parsedVersion
+
+		case "value":
+			// value must be a base64 encoded string
+			base64EncodedValue, err := evaluateLiteralExpr(value)
+			if err != nil {
+				return nil, errorf(value.Range(), "invalid base64 encoded value for parameterization of %q: %v", packageName, err)
+			}
+
+			decoded, err := base64.StdEncoding.DecodeString(base64EncodedValue)
+			if err != nil {
+				return nil, errorf(value.Range(), "invalid base64 encoded value for parameterization of %q: %v", packageName, err)
+			}
+
+			descriptor.Value = decoded
+		}
+	}
+
+	return descriptor, nil
+}
+
+// ReadPackageDescriptors reads the package blocks in the given file and returns a map of package names to the package.
+// The descriptors blocks are top-level program blocks have the following format:
+//
+//	package <name> {
+//	  baseProviderName = <name>
+//	  baseProviderVersion = <version>
+//	  baseProviderDownloadUrl = <url>
+//	  parameterization {
+//	      name = <name>
+//	      version = <version>
+//	      value = <base64 encoded string>
+//	  }
+//	}
+//
+// These package descriptors allow the binder loader to load package schema information from a package source.
+func ReadPackageDescriptors(file *syntax.File) (map[string]*schema.PackageDescriptor, hcl.Diagnostics) {
+	var diagnostics hcl.Diagnostics
+	packageDescriptors := map[string]*schema.PackageDescriptor{}
+	for _, node := range model.SourceOrderBody(file.Body) {
+		switch node := node.(type) {
+		case *hclsyntax.Block:
+			if node.Type != "package" {
+				// we only care about blocks of the form:
+				//    package <name> { ... }
+				continue
+			}
+
+			labels := node.Labels
+			if len(labels) != 1 {
+				diagnostics = append(diagnostics,
+					labelsErrorf(node, "package blocks must have exactly one label (the package name)"))
+				continue
+			}
+
+			packageName := labels[0]
+			// make sure we don't declare the same package twice
+			if _, ok := packageDescriptors[packageName]; ok {
+				diagnostics = append(diagnostics,
+					errorf(node.Range(), "package %q was already defined", packageName))
```

Comment:
Regardless of whether or not we do what I suggested above, I don't suppose we could tell users where the duplicate definitions are defined? Is this something we get for free/already do when binding other kinds of blocks?

Reply:
We don't get that for free. We only know which block we are binding now and report that it is a duplicate. We could keep track of the files from which the packages are defined to be able to report better errors